src/guidellm/scheduler/base.py

```diff
@@ -298,42 +294,81 @@ async def _run_sync(
     async def _run_async(
         self, benchmark: TextGenerationBenchmark, end_time: float, max_number: float
     ) -> AsyncGenerator[Union[TextGenerationResult, TextGenerationError], None]:
+        """
+        Notes:
+            if the Load Generation Mode is set to 'consistent' - timestamps should
+            not be generated in order to make as many requests as possible to
+            simulate concurrent clients interaction.
+        """
+
         tasks = []
         completed = 0
 
-        for index, (request, submit_at) in enumerate(
-            zip(self.generator, self.load_generator.times())
-        ):
-            while (index + 1 - completed) >= settings.max_concurrency:
-                await asyncio.sleep(0.1)
+        def _completed(_task: asyncio.Task) -> None:
+            nonlocal completed
+            completed += 1
+            _res = _task.result()
 
-            if index >= max_number or time.time() >= end_time or submit_at >= end_time:
-                break
+            if _res:
+                benchmark.request_completed(_res)
+                logger.debug("Request completed: {}", _res)
 
-            logger.debug(
-                "Running asynchronous request={} at submit_at={}",
-                request,
-                submit_at,
-            )
-
-            def _completed(_task: asyncio.Task) -> None:
-                nonlocal completed
-                completed += 1
-                _res = _task.result()
-
-                if _res:
-                    benchmark.request_completed(_res)
-                    logger.debug("Request completed: {}", _res)
+        if self.mode == "consistent":
+            if self.rate is None:
+                raise ValueError(
+                    "The `rate` must be specified in order to provide "
+                    "the concurrent execution"
+                )
+            for index, request in enumerate(self.generator):
+                while (index + 1 - completed) >= settings.max_concurrency:
```

Comment:
Shouldn't we ignore the max_concurrency setting in this case / raise an error if the concurrent requests number is smaller than the rate type we need to run at? 

Additionally, our check should be on the rate here and we would only allow a new request if we're below our rate or have passed our time / request count restrictions, right?